### PR TITLE
New merkle tree sc

### PR DESCRIPTION
**Pull request by anhdungle93**
Change current incremental merkle tree so that one can change content of each leaf, hence enables revocation

### CODE REVIEW

**Review by zwilling — CHANGES_REQUESTED** [review]
Nice implementation of the sparse merkle tree.
I added some comments on potential issues and code style.

**Review by zwilling — APPROVED** [review]
Only minor improvement suggestions. Do you work on them or merge it already?

File: tasks/createZKKYC.ts [hunk 1]
```diff
@@ -104,26 +104,26 @@ async function main(args: any, hre: HardhatRuntimeEnvironment) {
...
+  // find the smallest index of an empty list
+  let index = 0;
     leafIndices.sort();
     while (true) {
-        if (leafIndices.includes(index)) {
-            break;
-        } else {
-            index++;
-        }
+    if (!leafIndices.includes(index)) {
```
[the remark this answers is not in the document]
Comment (anhdungle93) [reply tasks/createZKKYC.ts]: added a new sorting

File: tasks/reissueZKKYC.ts [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+  // write output to file
+  const outputFileName = args.outputFile || `issuedZkKYCs/${zkKYC.leafHash}.json`;
+  fs.mkdirSync(path.dirname(outputFileName), { recursive: true });
+  fs.writeFileSync(outputFileName, JSON.stringify(output, null, 2));
+  console.log(chalk.green(`Written ZkKYC to output file ${outputFileName}`));
+
+  console.log(chalk.green("done"));
+}
+
+task("createZkKYC", "Task to create a zkKYC certificate with input parameters")
```
Comment (zwilling) [inline tasks/reissueZKKYC.ts:167]: I think the task name should be different from the other `createZkKYC` task.

**Pull request merged**